### ISSUE TITLE
On the future and API of CuPy

### ISSUE THREAD

**Issue by beam2d** [chainer/chainer#533]
There have been some discussions on the future and API of CuPy (#333, #520,  #522). I noticed that we have to give a clear direction to the development of CuPy. I discussed with other team members, and this is a report of the conclusion. Further discussion is welcome.

There are three points to care about: API policy, independence from this repository, and core features that we are planning to add in the future.
## TL;DR
- We want to allow CuPy to have extra interfaces compared to NumPy by some solid ways (allowing extra keywrod arguments and putting extra functions to special namespaces).
- We want to put CuPy to an isolated repository and package after adding high level kernel synthesis (still under design).
- For the future development, we aim at improving the performance of sequences of small operations (required for modern neural nets applications, in particular in NLP domain).
## API policy

The unique point of CuPy is its NumPy-compatible interface. We all agreed that we should keep this point. 100% compatibility is great, though I and @okuta san found that it is a very hard task and takes very long time to attain (and yet the feasibility is unclear).

We should keep in mind that CuPy is for Chainer in nature. Why we wrote this is that we want to make it easy to write fast code right now. I believe there must arise other GPU array libraries in the near future (e.g. DyND and libgpuarray). There are two choices for Chainer: continuing to use and maintain CuPy, or switching to such other libraries. Both futures are likely to realize, while we do not need to decide it yet. The important point is that we cannot switch to such libraries right now and maybe even for some years. Such other GPU array libraries might not be compatible with NumPy API, so the compatibility with NumPy must be a strong point of CuPy. On the other hand, if CuPy does not provide easy ways to write fast code compared to these libraries, then people might finally discard CuPy (it means they stop using Chainer) and use them in order to achieve higher performance. Since people use GPU for speed, we must think about the performance first. It might require extra interfaces compared to NumPy. It does not hurt the fact we can write CPU/GPU agnostic code by using common subset of NumPy/CuPy APIs.

In order to keep the interface clear, I want to add following restriction:
- Any functions can have extra keyword arguments.
- Extra functions should be put to dedicated namespaces, like `cupy.ext` or something.
  And also we might have to think about adding a wrapper of NumPy/CuPy APIs that provides a way to write CPU/GPU agnostic code on the extended interface.
## Independence from this repository

There are needs for making CuPy an independent repository or project (I mentioned this point in #333). We also agreed that CuPy must be an isolated package in the future, though it is not now.

Making CuPy an isolated package means it accepts usages outside of Chainer. It might greatly expand the user base, which accelerates the development. On the other hand, here again we have to think about the nature of CuPy that it is for Chainer. We are thinking about introducing function synthesis in the near future. It is like an optimization of a sequence of function applications and combines them into one function with fast synthesized kernels. It might require CuPy to support further kernel synthesis. In order to keep the cooperation of Chainer/CuPy development as much as possible during the development of this feature, I want to defer the isolation until the design is fixed at some level. Once the cooperation is done, we have to do the isolation as early as possible to accelerate the development. I think it is done within half a year.
## Core features that might be added in the future

Currently, we are thinking about following features:
- Porting to Cython: @okuta san is working on it, first in #499 to replace ctypes. We want to replace all Python code in CuPy to Cython. It is required to purely make CuPy much faster, in particular for usages that require a lot of small operations.
- High level kernel synthesis, which I wrote above. It is also for the performance on a lot of small operations. It also reduces the memory footprint of sequences of typical array computations.
- CUDA-stream support. It is required to leverage all the CUDA cores for small operations.
- Inter-process communication support, mainly for collaboration with multiprocessing package. It also enhances the parallel programming.

So the main subject is the speed up of sequences of small operations. This is mainly required for fast traning of recurrent nets, in particular for NLP domain. Since supporting simple and efficient code for recurrent nets and other dynamic neural nets is the primal goal of Chainer, making it fast in such usages is the largest demand on CuPy.

**Comment by bordingj** [comment]
Will custom kernel cached compilation still be supported ?
I write alot of custom kernels for example like this

	@cp.util.memoize(for_each_device=True)
	def _GetNoWhitespaces_kernel():
	    kernel_code = cp.carray.compile_with_cache("""
	    extern "C" __global__
	    void GetNoWhitespaces(const int* IDs, const int* whitespace_IDs, int* no_whitespaces,
		             const int no_whitespace_ids, const int T, const int N){
		    
		__shared__ int shared_vec[32];
		    
		if (blockIdx.y < N){
		    int no_whitespaces_i = 0;
		    int n; 
		    for (int t = threadIdx.x; t < T; t += 32){
		        n = t*N+blockIdx.y;
		        for (int k=0; k<no_whitespace_ids; k++){
		            if (IDs[n] == whitespace_IDs[k]){
		                no_whitespaces_i += 1;         
		                break;
		            }               
		        } 
		    }
		    shared_vec[threadIdx.x] = no_whitespaces_i;
		    __syncthreads();
		    
		    if (threadIdx.x == 0){
		        for (int j=0; j<32; j++){
		            no_whitespaces[blockIdx.y] += shared_vec[j];
		        }
		    }
		}
	    }
	    """)
	    return kernel_code.get_function('GetNoWhitespaces')

And then call them using

	def getNoWhitespaces(IDs, whitespace_IDs):
	    no_whitespace_ids = whitespace_IDs.shape[0]
	    T = IDs.shape[0]
	    N = IDs.shape[1]
	    no_whitespaces = cp.zeros((N,), dtype=np.int32)
	    
	    _GetNoWhitespaces = _GetNoWhitespaces_kernel()
	    
	    bdim, gdim = (32,1,1), (1,N,1)

	    _GetNoWhitespaces(grid=gdim, block=bdim,
		  args=(IDs, whitespace_IDs, no_whitespaces,
		        no_whitespace_ids, T, N)
		    )
	    return no_whitespaces

This is one of the reasons why I use chainer/cupy instead of Theano.

**Comment by tscohen** [comment]
Sounds good. A basic rule would be:
- It's ok if code using cupy does not work with numpy
- It's not ok if code that runs on numpy does not run on cupy (this is a bug or missing feature)

Are the extra functions that you want to add for speed any more general than what is in BLAS? BLAS already has many fused operations. Perhaps a way to stay CPU compatible would be to implement the scipy.linalg.blas interface.

**Comment by beam2d** [comment]
scipy.linalg.blas is a good starting point for CPU/GPU compatible support of efficient matrix operations. I actually do not want to make end users required to call them frequently, though. They are prone to embed a bug, and should be wrapped by well-tested libraries with more useful interface.

**Comment by beam2d** [comment]
@bordingj Yes, we should continue to support custom kernels with caching.

**Comment by beam2d** [comment]
Close since it's old. CuPy has already been made into an independent repository/package. Let's keep on developing cupy at https://github.com/cupy/cupy!
